### ISSUE TITLE
Model evaluation

### ISSUE THREAD

**Issue by Tandon-A** [Healthcare-Robotics/BodyPressure#10]
Hi Henry, 

I was trying to understand the model evaluation code for metric calculation and had some doubts regarding that. 

1. What is the use of this transformation in [lib_pyrender_depth](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_depth.py#L202-L203)?

2. Is [calculate_results.py](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/networks/calculate_results.py) the main file which you use to obtain the numbers presented in the paper? Are you calling this file separately?  
Not able to understand [this](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/networks/calculate_results.py#L213) part in v2v pressure error calculation. 

3.  Why is the normalized avg area multiplied by the shape of area avg reduced? [Line](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L187) 

Best, 
Abhishek

**Comment by Tandon-A** [comment]
Also, the method [get_triangle_area_vert_weight](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L134), as I understand, is supposed to output the average area of the triangles a vertex is connected to for each vertex. 

[At line 158](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L158) you are adding each area for that vertex in the list and then at line 161 you are taking a mean of those areas as expected. 

But I am not able to understand the part of dividing by the sum of the area of all vertices at [line 186.](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L186).

**Comment by henryclever** [comment]
Hi, these are very good questions! 

At the moment I am buried in work and will need some time to look into them. I'll try to get to them this week -- if not, this weekend.

-Henry

**Comment by henryclever** [comment]
_What is the use of this transformation in [lib_pyrender_depth](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_depth.py#L202-L203)?_
This transformation equates to a 1.0m lateral shift of the pressure projection rendering from the smpl wireframe+flat pressure mat rendering. This is defined in the get_transforms function. You can see it in the SLP-3Dfits dataset visualization section of the README. The point cloud+green mesh rendering is 1.0m in the other direction. This is purely for visualization purpose.

_Is [calculate_results.py](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/networks/calculate_results.py) the main file which you use to obtain the numbers presented in the paper? Are you calling [this](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/networks/calculate_results.py#L213) file separately?_
Yes. However, it does not actually do any inference – it takes as input a pickle file containing predictions and ground truth that is generated in the previous step (L253-255).
_Not able to understand this part in v2v pressure error calculation._
-The code at line 213 was used to compute a distribution of pressure magnitudes for the vertices in the real data. 6890 vertices per pose, 45 poses per participant. I at one point wanted to understand the frequency of pressure magnitudes at different steps 0-10-20... mmHg etc, but I don’t think I used it anywhere in the paper.

I’ll follow up again regarding the pressure computation  / vertex weighting.

**Comment by henryclever** [comment]
_But I am not able to understand the part of dividing by the sum of the area of all vertices at [line 186.](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L186)._

The purpose of this is to limit the influence of vertex weighting to only vertices where the surface of the body could contact the bed ... i.e. any part of the body that is facing slightly upward should have zero pressure. This is ensured in L208-224 here: https://github.com/Healthcare-Robotics/BodyPressure/blob/5962b31d89483c0449220960b0a27c2044229f19/lib_py/lib_pyrender_depth.py . The variable `vertex_pressure_init_list_EST` contains a 6890 array of unweighted vertex pressures where roughly half of them are set to zero from parts of the body facing upwards. Thus it should be multiplied by a weighting array that is of size 6890, but the weighting should only be computed among vertices where pressure is can be greater than zero from the body surface pointing down (~3500 vertices). The weighting for all downfacing vertex pressures should sum to 1 * Ndownfacingverts. So without the adjustment on L186 and L187, the weighting array for those vertices will not sum to 1 *Ndownfacingverts.

Alternatively I could have reduced the entire size of the vertex pressure array to make it about half of 6890, which would obviate the need for L186 and L187, but that requires changing the size of the array dynamically each computation.... much easier to mask and adjust the weighting by the fraction of downfacing pressure vertices.

**Comment by Tandon-A** [comment]
Thanks Henry for the detailed answers.

**Comment by Tandon-A** [comment]
One more doubt. Why are you not adding the area from the last face? [L157](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L157) (The if condition after 'and' in this line). 

I think due to this the [avg area](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L180) is coming out wrong.

**Comment by Tandon-A** [comment]
I prepared a script trying to check the function. Due to that if condition checking for the last index the script does not calculate the correct avg for the last face index. 

```python
import torch 
import numpy as np
import math

faces = torch.tensor([0, 0, 1, 1, 2, 2]).view(6, 1) # faces for the sample case
areas = torch.rand(6, 1) #define some random areas

def get_area_norm_torch(faces, areas):
    b = faces.repeat(1, 3)
    d = areas.repeat(1, 3)

    e = torch.arange(3).view(1, 3).repeat(6, 1)
    f = torch.where(b == e, 1, 0)

    area_avg = torch.nan_to_num(torch.sum(f*d, dim=0)/torch.sum(f, dim=0))
    return area_avg


def get_area_norm_np_edited(faces, areas):
    faces_sorted = faces.numpy()[:, 0]
    A_sorted = areas.numpy()[:, 0]
    last_face = 0
    area_minilist = []
    area_avg_list = []
    face_sort_list = [] # take the average area for all the trianges surrounding each vert
    for vtx_connect_idx in range(np.shape(faces_sorted)[0]):
        if faces_sorted[vtx_connect_idx] == last_face:
            area_minilist.append(A_sorted[vtx_connect_idx])
        elif faces_sorted[vtx_connect_idx] > last_face or vtx_connect_idx == np.shape(faces_sorted)[0]-1:
            if len(area_minilist) != 0:
                area_avg_list.append(np.mean(area_minilist))
            else:
                area_avg_list.append(0)
            face_sort_list.append(last_face)
            area_minilist = []
            last_face += 1
            if faces_sorted[vtx_connect_idx] == last_face:
                area_minilist.append(A_sorted[vtx_connect_idx])
            elif faces_sorted[vtx_connect_idx] > last_face:
                num_tack_on = np.copy(faces_sorted[vtx_connect_idx] - last_face)
                for i in range(num_tack_on):
                    area_avg_list.append(0)
                    face_sort_list.append(last_face)
                    last_face += 1
                    if faces_sorted[vtx_connect_idx] == last_face:
                        area_minilist.append(A_sorted[vtx_connect_idx])
    if len(area_minilist) != 0:
        area_avg_list.append(np.mean(area_minilist))    

    #print np.mean(area_avg_list), 'area avg'

    area_avg = np.array(area_avg_list)
    # print (area_avg)


    area_avg_red = area_avg[area_avg > 0] #find out how many of the areas correspond to verts facing the camera

    norm_area_avg = area_avg/np.sum(area_avg_red)
    norm_area_avg = norm_area_avg*np.shape(area_avg_red)

    # print (area_avg_red.shape)
    # print (norm_area_avg)
    # print (sum(norm_area_avg))
    return area_avg

# Code from [get_triangle_area_weight](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L156-L180)
def get_area_norm_np(faces, areas):
    faces_sorted = faces.numpy()[:, 0]
    A_sorted = areas.numpy()[:, 0]
    last_face = 0
    area_minilist = []
    area_avg_list = []
    face_sort_list = [] # take the average area for all the trianges surrounding each vert
    for vtx_connect_idx in range(np.shape(faces_sorted)[0]):
        if faces_sorted[vtx_connect_idx] == last_face and vtx_connect_idx != np.shape(faces_sorted)[0]-1:
            area_minilist.append(A_sorted[vtx_connect_idx])
        elif faces_sorted[vtx_connect_idx] > last_face or vtx_connect_idx == np.shape(faces_sorted)[0]-1:
            if len(area_minilist) != 0:
                area_avg_list.append(np.mean(area_minilist))
            else:
                area_avg_list.append(0)
            face_sort_list.append(last_face)
            area_minilist = []
            last_face += 1
            if faces_sorted[vtx_connect_idx] == last_face:
                area_minilist.append(A_sorted[vtx_connect_idx])
            elif faces_sorted[vtx_connect_idx] > last_face:
                num_tack_on = np.copy(faces_sorted[vtx_connect_idx] - last_face)
                for i in range(num_tack_on):
                    area_avg_list.append(0)
                    face_sort_list.append(last_face)
                    last_face += 1
                    if faces_sorted[vtx_connect_idx] == last_face:
                        area_minilist.append(A_sorted[vtx_connect_idx])

    #print np.mean(area_avg_list), 'area avg'

    area_avg = np.array(area_avg_list)
    # print (area_avg)


    area_avg_red = area_avg[area_avg > 0] #find out how many of the areas correspond to verts facing the camera

    norm_area_avg = area_avg/np.sum(area_avg_red)
    norm_area_avg = norm_area_avg*np.shape(area_avg_red)

    # print (area_avg_red.shape)
    # print (norm_area_avg)
    # print (sum(norm_area_avg))
    return area_avg


areas_torch = get_area_norm_torch(faces, areas)
areas_numpy = get_area_norm_np(faces, areas)
areas_numpy_edited = get_area_norm_np_edited(faces, areas)

print ('faces', faces.reshape(1, 6))
print ('areas', areas.reshape(1, 6))
print ('area avg torch', areas_torch)
print ('area avg numpy', areas_numpy)
print ('area avg numpy edited', areas_numpy_edited)
```

This sets the faces and defines random area values for the vertices. Then I run three functions, get_area_norm_torch to get the avg area through a torch code, get_area_norm_np ([referring your code](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/lib_pyrender_functions.py#L156-L180)), and edited version of your code. 

If you run this, the output of the last face in area avg numpy is not correct. 

Please do let me know if I missed something.

**Comment by henryclever** [comment]
If I am not mistaken, the "last" face occurring when `vtx_connect_idx == np.shape(faces_sorted)[0]-1:` is only `True` once for the entire body, correct? How large is the discrepancy between the avg areas between your edited code and mine? 

I thought I had some reason for doing this (maybe something specific to the SMPL mesh?) ... hard to remember at the moment.

-Henry
